File: s/src/com/novacloud/novaone/expression/expFunction/ExpCommon.java
package com.novacloud.novaone.expression.expFunction;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Date;

import com.novacloud.novaone.common.JSONProcessor;

import net.sf.json.JSONObject;

//通用方法集合
public class ExpCommon { 
	//连接
	public String andString(Object param1, Object param2){
		String p1 = param1 == null ? "" : param1.toString();
		String p2 = param1 == null ? "" : param2.toString(); 
		return p1 + p2;
	}
	//加法
	public BigDecimal add(BigDecimal param1, BigDecimal param2){
		if(param1 == null){
			return param2;
		}
		else if(param2 == null){
			return param1;
		}
		else{
			return param1.add(param2);
		}
	}
	//加法
	public String add(String param1, String param2){
		String p1 = param1 == null ? "" : param1.toString();
		String p2 = param1 == null ? "" : param2.toString(); 
		return p1 + p2;
	}
	
	//正数
	public BigDecimal add(BigDecimal param1){
		return param1;
	}

	//负数
	public BigDecimal subtract(BigDecimal param1){
		if(param1 == null){
			return null;
		} 
		else{
			return BigDecimal.ZERO.subtract(param1);
		}
	}

	//减法
	public BigDecimal subtract(BigDecimal param1, BigDecimal param2){
		if(param2 == null){
			return param1;
		}
		else if(param1 == null){
			return BigDecimal.ZERO.subtract(param1);
		}
		else{
			return param1.subtract(param2);
		}
	}
	
	//减法
	public BigDecimal multiply(BigDecimal param1, BigDecimal param2){
		if(param1 == null || param2 ==null){
			return null;
		}
		else{
			return param1.multiply(param2);
		}
	}
	
	//除法
	public BigDecimal divide(BigDecimal param1, BigDecimal param2) throws Exception{
		if(param2 == null){
			throw new Exception("除数不能为空");
		}
		else if(param1 == null){
			return null;
		}
		else{
			return param1.divide(param2, 20, RoundingMode.HALF_DOWN);
		}
	}
	
	//相等
	public boolean equal(BigDecimal param1, BigDecimal param2){
		if(param1 == null){
			return param2 == null;
		}
		else{
			return param1.equals(param2);
		}
	} 
	
	//相等
	public boolean equal(String param1, String param2){
		if(param1 == null){
			return param2 == null;
		}
		else{
			return param1.equals(param2);
		}
	} 
	
	//相等
	public boolean equal(Date param1, Date param2){
		if(param1 == null){
			return param2 == null;
		}
		else{
			return param1.equals(param2);
		}
	} 
	
	//相等
	public boolean equal(Object param1, Object param2){ 
		if(param1 == null){
			return param2 == null;
		}
		else{
			return param1.equals(param2);
		}
	} 
	
	//大于
	public boolean moreThan(BigDecimal param1, BigDecimal param2){
		if(param1 == null || param2 == null){
			throw new RuntimeException("值为空，无法比较大小");
		}
		else{
			return param1.compareTo(param2) == 1;
		}
	}   
	
	//小于
	public boolean lessThan(BigDecimal param1, BigDecimal param2){
		if(param1 == null || param2 == null){
			throw new RuntimeException("值为空，无法比较大小");
		}
		else{
			return param1.compareTo(param2) == -1;
		}
	}  
	
	//大于等于
	public boolean moreThanOrEqual(BigDecimal param1, BigDecimal param2){
		if(param1 == null || param2 == null){
			throw new RuntimeException("值为空，无法比较大小");
		}
		else{
			return param1.compareTo(param2) >= 0;
		}
	}  
	
	//小于等于
	public boolean lessThanOrEqual(BigDecimal param1, BigDecimal param2){
		if(param1 == null || param2 == null){
			throw new RuntimeException("值为空，无法比较大小");
		}
		else{
			return param1.compareTo(param2) <= 0;
		}
	}  

	//Json转字符串
	public String jsonToString(JSONObject json){
		String s = JSONProcessor.jsonToStr(json);
		return s;
	}
	
	//字符串转Json
	public JSONObject stringToJson(String str) throws Exception{
		JSONObject obj = JSONProcessor.strToJSON(str);
		return obj;
	}

	//IIF判断
	public String iif(boolean check, String returnValue1, String returnValue2){
		 return check ? returnValue1 : returnValue2;
	}
	//IIF判断
	public Date iif(boolean check, Date returnValue1, Date returnValue2){
		 return check ? returnValue1 : returnValue2;
	}
	//IIF判断
	public BigDecimal iif(boolean check, BigDecimal returnValue1, BigDecimal returnValue2){
		 return check ? returnValue1 : returnValue2;
	}
	//IIF判断
	public JSONObject iif(boolean check, JSONObject returnValue1, JSONObject returnValue2){
		 return check ? returnValue1 : returnValue2;
	}
	
	//and
	public boolean and(boolean param1, boolean param2){
		 return param1 && param2;
	}
	//or
	public boolean or(boolean param1, boolean param2){
		 return param1 || param2;
	}
}

File: s/WebContent/novaone-style/plugins/novaone/expression/expCommon.js
//对话框
var expCommon = {
	andString: function(var1, var2){
		var s = "";
		if(var1 != null){
			s += var1;
		}
		if(var2 != null){
			s += var2;
		}
		return s;
	},
	//加减乘除
	add:function(var1, var2){
		if(var1 == null){
			var1 = 0;
		}
		else if(cmnPcr.isDecimal(var1)){
			var1 = parseFloat(var1); 
		}
		else{
			throw new Error(var1.toString() + "不是数值");
		}
		if(var2 == null){
			var2 = 0;
		}
		else if(cmnPcr.isDecimal(var2)){
			var2 = parseFloat(var2);
		}
		else{
			throw new Error(var2.toString() + "不是数值");
		}
		return cmnPcr.floatAdd(var1, var2);
		/*
		if(var1 == null){
			return var2;
		}
		else if(var2 == null){
			var var1;
		}	 
		else if(typeof var1 === "number" && typeof var2 === "number"){
			 return cmnPcr.floatAdd(var1, var2);
		}
		else {
			 return var1 + var2;
		}*/
	 },
	 subtract:function(var1, var2){
		 if(var2 == undefined){
			 return 0 - var1;
		 }
		 else{
			 return cmnPcr.floatAdd(var1, 0 - var2); 
		 }
	 },
	 multiply:function(var1, var2){
	    var m = 0, s1 = var1.toString(), s2 = var2.toString();  
	    try{ m += s1.split(".")[1].length; }catch(e){}  
	    try{ m += s2.split(".")[1].length; }catch(e){}  
	    return Number(s1.replace(".",""))*Number(s2.replace(".",""))/Math.pow(10,m)  
		 //return var1 * var2;
	 },
	 divide:function(var1, var2){
		 try{
		 	if(var1 == null){
		 		throw new Error("被除数不能为空");
		 	}
		 	else if(var2 == null){
		 		throw new Error("除数不能为空");
		 	}
		 	else if(var2 == 0){
				throw new Error("除数不能为0");
		 	}
		 	else {
		 	   var t1 = 0, t2 = 0, r1, r2;  
			    try{t1 = var1.toString().split(".")[1].length}catch(e){}  
			    try{t2 = var2.toString().split(".")[1].length}catch(e){}  
			    with(Math){  
			        r1=Number(var1.toString().replace(".",""))  
			        r2=Number(var2.toString().replace(".",""))  
			        return (r1/r2)*pow(10,t2-t1);  
			    }	 	
				//return var1 / var2;
			}
		}
		catch(ex){
			throw ex;
		}
	 },
	 
	 //判断相等
	 equal:function(var1, var2){
		 return var1 == var2;
	 },

	 //大于
	 moreThan:function(var1, var2){
		 return var1 > var2;
	 },
	 //小于
	 lessThan:function(var1, var2){
		 return var1 < var2;
	 },
	 //大于等于
	 moreThanOrEqual:function(var1, var2){
		 return var1 >= var2;
	 },
	 //小于等于
	 lessThanOrEqual:function(var1, var2){
		 return var1 <= var2;
	 },
	 
	 //字符串和json对象互转
	 jsonToString:function(json){
		 return cmnPcr.jsonToStr(json);
	 },
	 stringToJson:function(str){
		 return cmnPcr.strToJson(str);
	 },
	 
	 //IIF
	 iif:function(boolValue, returnValue1, returnValue2){
		 return boolValue ? returnValue1 : returnValue2;
	 },
	 
	 //and
	 and:function(param1, param2){
		 return param1 && param2;
	 },
	 //or
	 or:function(param1, param2){
		 return param1 || param2;
	 }
}